core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -0,0 +1,174 @@
+import org.apache.kafka.common.TopicPartition
+
+import scala.annotation.tailrec
+import scala.collection.immutable
+import scala.util.{Failure, Success, Try}
+
+private[kafka] abstract class SubSourceLogic[K, V, Msg](
+    shape: SourceShape[(TopicPartition, Source[Msg, NotUsed])],
+    settings: ConsumerSettings[K, V],
+    subscription: AutoSubscription
+) extends GraphStageLogic(shape) with PromiseControl with MessageBuilder[K, V, Msg] {
+  var consumer: KafkaAsyncConsumer[K, V] = _
+  var buffer: immutable.Queue[TopicPartition] = immutable.Queue.empty
+  var subSources: Map[TopicPartition, Control] = immutable.Map.empty
+  implicit lazy val ec = ActorMaterializer.downcast(materializer).executionContext
+
+  override def preStart(): Unit = {
+    super.preStart()
+    consumer = KafkaAsyncConsumer[K, V](settings)(ActorMaterializer.downcast(materializer).system)
+    subscription match {
+      case TopicSubscription(topics) =>
+        consumer.subscribe(topics, KafkaAsyncConsumer.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+      case TopicSubscriptionPattern(topics) =>
+        consumer.subscribe(topics, KafkaAsyncConsumer.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+    }
+  }
+
+  val partitionAssignedCB = getAsyncCallback[Iterable[TopicPartition]] { tps =>
+    buffer = buffer.enqueue(tps.toList)
+    pump()
+  }
+  val partitionRevokedCB = getAsyncCallback[Iterable[TopicPartition]] { r =>
+    r.map(subSources.get).foreach(_.foreach(_.shutdown()))
+    subSources --= r
+  }
+
+  val subsourceCancelledCB = getAsyncCallback[TopicPartition] { tp =>
+    subSources -= tp
+    buffer :+= tp
+    pump()
+  }
+
+  val subsourceStartedCB = getAsyncCallback[(TopicPartition, Control)] {
+    case (tp, control) => subSources += (tp -> control)
+  }
+
+  setHandler(shape.out, new OutHandler {
+    override def onPull(): Unit = pump()
+  })
+
+  def createSource(tp: TopicPartition): Source[Msg, NotUsed] = {
+    Source.fromGraph(new SubSourceStage(tp, consumer))
+  }
+
+  @tailrec
+  private def pump(): Unit = {
+    if (buffer.nonEmpty && isAvailable(shape.out)) {
+      val (tp, remains) = buffer.dequeue
+      buffer = remains
+      push(shape.out, (tp, createSource(tp)))
+      pump()
+    }
+  }
+
+  override def postStop(): Unit = {
+    subSources.foreach {
+      case (_, control) => control.shutdown()
+    }
+    consumer.close().onSuccess { case _ => shutdowned() }
+    super.postStop()
+  }
+
+  val stopCB = getAsyncCallback[Unit]({ _ =>
+    subSources.foreach {
+      case (_, control) => control.stop()
+    }
+    completeStage()
+  })
+
+  val performStop = getAsyncCallback[Unit]({ _ =>
+    complete(shape.out)
+    stopped()
+  }).invoke _
+
+  val performShutdown = getAsyncCallback[Unit]({ _ =>
+    completeStage()
+  }).invoke _
+
+  class SubSourceStage(tp: TopicPartition, consumer: KafkaAsyncConsumer[K, V]) extends GraphStage[SourceShape[Msg]] {
```

Comment:
This substream have no `Control` because you can not decline subscription if it was assigned by kafka broker